### PR TITLE
lwm2m: init messages, pendings and replies on lwm2m_engine_start

### PR DESCRIPTION
**Pull request by retfie**
On network with poor coverage, very often request/observe packets doesn't get it's ACK and consumes from pendings/replies/message stacks. In such cases when LWM2M engine tries to recover by resetting its state, it fails because of lack of free messages.

Signed-off-by: Kiril Petrov <retfie@gmail.com>

### CODE REVIEW

**Comment by mike-scott** [comment]
This should be split into 2 commits.  First to add the COAP functions.  Second to implement them in the LwM2M code.

NOTE: I haven't really reviewed this yet.

File: include/net/coap.h [hunk 1]
```diff
@@ -702,6 +702,9 @@ struct coap_pending *coap_pending_next_unused(
 struct coap_reply *coap_reply_next_unused(
 	struct coap_reply *replies, size_t len);
 
+void coap_pendings_clear(struct coap_pending *pendings, size_t len);
```
Comment (rlubos) [inline include/net/coap.h:705]: Would be good to add doxygen docs for these functions, just as other coap functions have.

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 2]
```diff
@@ -4006,6 +4016,11 @@ int lwm2m_engine_context_close(struct lwm2m_ctx *client_ctx)
 
 void lwm2m_engine_context_init(struct lwm2m_ctx *client_ctx)
 {
+	messages_clear(client_ctx);
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:4019]: It's more of a cleanup than initialization, IMO it would fit better inside `lwm2m_engine_context_close`

**Review by rlubos — COMMENTED** [review]
I've also got reports that coap pendings are leaking, so that after some time, when LWM2M engine tries to register, it is not able to operate anymore. I'm still investigating, what is the root cause for this behavior, but a general cleanup, when lwm2m context is closed sounds like a good idea.

**Comment by retfie** [comment]
Looks like there is some issue with OpenThread stack again (devices can't attach to OT network) , and I couldn't verify them on top of tree, so I cherry-pick them on my last know working OpenThread tag to be able to check LWM2M stack from end-to-end. I'll let some of my devices to work for several days before mark these 2 changes as stable enough.

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 3]
```diff
@@ -4008,6 +4014,10 @@ int lwm2m_engine_context_close(struct lwm2m_ctx *client_ctx)
 
 void lwm2m_engine_context_init(struct lwm2m_ctx *client_ctx)
 {
+	coap_pendings_clear(client_ctx->pendings,
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:4017]: I've meant those as well, I'd see them in `lwm2m_engine_context_close`, no point splitting cleanup operations.

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 4]
```diff
@@ -3997,6 +4000,9 @@ int lwm2m_engine_context_close(struct lwm2m_ctx *client_ctx)
 		}
 	}
 
+	for (i = 0, msg = messages; i < len; i++, msg++)
+		memset(msg, 0, sizeof(struct lwm2m_message));
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:4004]: nit: Zephyr requires brackets even around single-line blocks. Same in new coap functions.

**Review by rlubos — COMMENTED** [review]
I'd appreciate if we could move it out of draft soon-enough.

**Review by rlubos — APPROVED** [review]

**Review by jukkar — CHANGES_REQUESTED** [review]
I would like to see the commit subjects changed a bit, so please add `net: ` prefix to them. So `net: coap: ...` and `net: lwm2m: ...`
Other than that, looks ok.

**Comment by mike-scott** [comment]
Testing this and @rlubos PR https://github.com/zephyrproject-rtos/zephyr/pull/24668 today

File: include/net/coap.h [hunk 5]
```diff
@@ -767,6 +767,15 @@ bool coap_pending_cycle(struct coap_pending *pending);
  */
 void coap_pending_clear(struct coap_pending *pending);
 
+/**
+ * @brief Cancels all pending retransmissions, so they becomes
```
Comment (mike-scott) [inline include/net/coap.h:771]: s/becomes/become/g

File: include/net/coap.h [hunk 6]
```diff
@@ -767,6 +767,15 @@ bool coap_pending_cycle(struct coap_pending *pending);
...
+ * available again.
+ *
+ * @param pendings Array of pendings to be canceled.
```
Comment (mike-scott) [inline include/net/coap.h:774]: To be consistent with other API descriptions above let's use this:
```
 * @param pendings Pointer to the array of #coap_pending structures
 * @param len Size of the array of #coap_pending structures
```

File: include/net/coap.h [hunk 7]
```diff
@@ -775,6 +784,14 @@ void coap_pending_clear(struct coap_pending *pending);
  */
 void coap_reply_clear(struct coap_reply *reply);
 
+/**
+ * @brief Cancels all replies, so they becomes available again.
```
Comment (mike-scott) [inline include/net/coap.h:788]: s/becomes/become/g

File: include/net/coap.h [hunk 8]
```diff
@@ -775,6 +784,14 @@ void coap_pending_clear(struct coap_pending *pending);
...
+ *
+ * @param replies Array of replies to be canceled
+ * @param len Number of replies into array
```
Comment (mike-scott) [inline include/net/coap.h:791]: Here let's use these descriptions to be consistent with other API descriptions:
```
 * @param replies Pointer to the array of #coap_reply structures
 * @param len Size of the array of #coap_reply structures
```

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 9]
```diff
@@ -3997,6 +4000,15 @@ int lwm2m_engine_context_close(struct lwm2m_ctx *client_ctx)
 		}
 	}
 
+	for (i = 0, msg = messages; i < len; i++, msg++) {
+		memset(msg, 0, sizeof(struct lwm2m_message));
+	}
+
+	coap_pendings_clear(client_ctx->pendings,
+			CONFIG_LWM2M_ENGINE_MAX_PENDING);
```
Comment (mike-scott) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:4008]: Can we line up this line with `client_ctx->pendings`?

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 10]
```diff
@@ -3997,6 +4000,15 @@ int lwm2m_engine_context_close(struct lwm2m_ctx *client_ctx)
...
+	coap_replies_clear(client_ctx->replies,
+			CONFIG_LWM2M_ENGINE_MAX_REPLIES);
```
Comment (mike-scott) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:4010]: Can we line up this line with `client_ctx->replies`?

**Review by mike-scott — CHANGES_REQUESTED** [review]
Formatting / spelling / comment nits.

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 11]
```diff
@@ -3981,6 +3981,9 @@ int lwm2m_engine_context_close(struct lwm2m_ctx *client_ctx)
 	struct observe_node *obs, *tmp;
 	sys_snode_t *prev_node = NULL;
 	int sock_fd = client_ctx->sock_fd;
+	struct lwm2m_message *msg;
+	size_t len = CONFIG_LWM2M_ENGINE_MAX_MESSAGES;
```
Comment (mike-scott) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:3985]: Minor nit: we can drop the `len` variable here

and change the `for` loop that uses it to this below:
```
	for (i = 0, msg = messages; i < CONFIG_LWM2M_ENGINE_MAX_MESSAGES;
	     i++, msg++) {
```

**Comment by mike-scott** [comment]
@retfie Thank you for submitting this PR.  It's a nice cleanup.  Once my review comments are addressed, I don't see anything else that would hold this up from being merged.

File: subsys/net/lib/lwm2m/lwm2m_engine.c (hunk 9 above)
Comment (mike-scott) [reply subsys/net/lib/lwm2m/lwm2m_engine.c]: @retfie Unfortunately, moving the CONFIG_LWM2M_ENGINE_MAX_PENDING up to the line above has broken the line length rule for Zephyr :/

File: subsys/net/lib/lwm2m/lwm2m_engine.c (hunk 10 above)
Comment (mike-scott) [reply subsys/net/lib/lwm2m/lwm2m_engine.c]: @retfie Unfortunately, moving the CONFIG_LWM2M_ENGINE_MAX_REPLIES up to the line above has broken the line length rule for Zephyr.

I was imagining something like this:
```
	coap_replies_clear(client_ctx->replies,
			   CONFIG_LWM2M_ENGINE_MAX_REPLIES);
```

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 12]
```diff
@@ -3997,6 +3999,14 @@ int lwm2m_engine_context_close(struct lwm2m_ctx *client_ctx)
 		}
 	}
 
+	for (i = 0, msg = messages; i < CONFIG_LWM2M_ENGINE_MAX_MESSAGES;
+	     i++, msg++) {
+		memset(msg, 0, sizeof(struct lwm2m_message));
+	}
+
+	coap_pendings_clear(client_ctx->pendings, CONFIG_LWM2M_ENGINE_MAX_PENDING);
```
Comment (mike-scott) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:4007]: @retfie Unfortunately, moving the CONFIG_LWM2M_ENGINE_MAX_PENDING up to the line above has broken the line length rule for Zephyr :/

I was imagining something like this:
```
	coap_pendings_clear(client_ctx->pendings,
			    CONFIG_LWM2M_ENGINE_MAX_PENDING);
```

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 13]
```diff
@@ -3997,6 +3999,14 @@ int lwm2m_engine_context_close(struct lwm2m_ctx *client_ctx)
...
+	coap_replies_clear(client_ctx->replies, CONFIG_LWM2M_ENGINE_MAX_REPLIES);
```
Comment (mike-scott) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:4008]: @retfie Unfortunately, moving the CONFIG_LWM2M_ENGINE_MAX_REPLIES up to the line above has broken the line length rule for Zephyr.

I was imagining something like this:
```
	coap_replies_clear(client_ctx->replies,
			   CONFIG_LWM2M_ENGINE_MAX_REPLIES);
```

**Review by mike-scott — CHANGES_REQUESTED** [review]
Needs a minor tweak to pass formatting checks

File: subsys/net/lib/lwm2m/lwm2m_engine.c (hunk 12 above)
Comment (retfie) [reply subsys/net/lib/lwm2m/lwm2m_engine.c]: @mike-scott  sorry, short tab to blame :/ . Fixed.

**Comment by jukkar** [comment]
> I would like to see the commit subjects changed a bit, so please add `net: ` prefix to them. So `net: coap: ...` and `net: lwm2m: ...`

I think you missed my comment about the commit subjects.

**Comment by retfie** [comment]
> > I would like to see the commit subjects changed a bit, so please add `net: ` prefix to them. So `net: coap: ...` and `net: lwm2m: ...`
> 
> I think you missed my comment about the commit subjects.
It was there, but somehow it was dropped at some of rebases. Adding it back.

**Review by jukkar — APPROVED** [review]
LGTM

**Review by mike-scott — APPROVED** [review]
LGTM.  Thanks!

**Pull request merged**